File: golang/maximum_subarray.go
func maxSubArray(nums []int) int {
	dp := make([]int, len(nums))
	dp[0] = nums[0]
	for i := 1; i < len(nums); i++ {
		// choice to either include the prev max, or keep current num
		dp[i] = max(nums[i], dp[i-1]+nums[i])
	}

	sum := dp[0]
	for i := 1; i < len(dp); i++ {
		if dp[i] > sum {
			sum = dp[i]
		}
	}
	return sum
}
func max(x int, y int) int {
	if x > y {
		return x
	}
	return y
}

File: typescript/maximum_subarray.ts
function maxSubArray(nums: number[]): number {
	// max subarray implies contiguous sum so choice is either start anew or sum with existing

	let dp = new Array(nums.length)
	dp[0] = nums[0]

	let largest = dp[0]

	for (let i = 1; i < dp.length; i++) {
		dp[i] = Math.max(nums[i], nums[i] + dp[i - 1])

		largest = Math.max(largest, dp[i])
	}

	return largest
}

// don't need array declaration bc contiguous
function maxSubArray(nums: number[]): number {
	let prev = 0
	let max = -(10 ** 4)

	for (let i = 0; i < nums.length; i++) {
		prev = Math.max(prev + nums[i], nums[i])
		max = Math.max(max, prev)
	}

	return max
}
